runtime/vm/resolvesupport.cpp

```diff
@@ -1851,6 +1851,131 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
 
 }
 
+j9object_t
+resolveConstantDynamic(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpIndex, UDATA resolveFlags)
+{
+	J9RAMConstantDynamicRef *ramCPEntry = (J9RAMConstantDynamicRef*)ramCP + cpIndex;
+	j9object_t value = NULL;
+	J9JavaVM *vm = vmThread->javaVM;
+	bool resolved = false;
+
+RETRY:
+	/* Fast path resolution for previously resolved entries */
+	if (NULL != ramCPEntry->value) {
+		value = ramCPEntry->value;
+		resolved = true;
+	} else {
+		/* check if already resolved to NULL or exception */
+		if (NULL != ramCPEntry->exception) {
+			J9Class *throwable = J9VMJAVALANGTHROWABLE_OR_NULL(vm);
+			J9Class *clazz = J9OBJECT_CLAZZ(vmThread, ramCPEntry->exception);
+			if (ramCPEntry->exception == vm->voidReflectClass->classObject) {
+				/* Resolved with null reference */
+				resolved = true;
+			} else if (isSameOrSuperClassOf(throwable, clazz)) {
+				/* Resolved with exception */
+				VM_VMHelpers::setExceptionPending(vmThread, ramCPEntry->exception);
+				resolved = true;
+			}
+		}
+	}
+
+	if (!resolved) {
+		/* Slow path, attempt to acquire resolve mutex or retry, or wait */
+		omrthread_monitor_enter(vm->constantDynamicMutex);
+		bool retry = false;
+		if (NULL != ramCPEntry->value) {
+			retry = true;
+		} else {
+			/* check if already resolved to NULL or exception */
+			if (NULL != ramCPEntry->exception) {
+				J9Class *throwable = J9VMJAVALANGTHROWABLE_OR_NULL(vm);
+				J9Class *clazz = J9OBJECT_CLAZZ(vmThread, ramCPEntry->exception);
+				if (ramCPEntry->exception == vm->voidReflectClass->classObject) {
+					/* Resolved with null reference */
+					retry = true;
+				} else if (isSameOrSuperClassOf(throwable, clazz)) {
+					/* Resolved with exception */
+					retry = true;
+				} else if (ramCPEntry->exception != vmThread->threadObject) {
+					/* Another thread is currently resolving this entry */
+					// internalReleaseVMAccess(vmThread);
+					omrthread_monitor_wait(vm->constantDynamicMutex);
+					omrthread_monitor_exit(vm->constantDynamicMutex);
```

Comment:
This will end up being exitting twice (below in the retry path).  Might be better to just exit and goto inline instead.

Reply:
I believe this exit is something I forgot to remove from last update, there are multiple retry paths, inlining all would make the code look bad. I ll remove this exit

Reply:
You can't remove this one - you must exit the monitor before reacquiring VM access.

Reply:
Right, I see what you meant
Corrected